docs/rfcs/sync-manager.md

```diff
@@ -0,0 +1,171 @@
+        config: GlobalConfig,
+
+        // The token server token.
+        storage_token: Vec[u8] (??), // not clear on the type here, but whatever.
+
+        // Info about the collections managed by this engine. In most cases there
+        // will be exactly 1.
+        collections: HashMap<String, CollectionState>
+    }
+
+    // The state for a collection. Most engines will manage a single collection, but
+    // there may be many.
+    struct CollectionState {
+        // The info/collections response for this collection.
+        collection_info: CollectionInfo,
+
+        // The current syncID for the engine, as read from meta/global. May be None.
+        // Engines are responsible for persisting this and taking action when it differs.
+        sync_id: Option<SyncGuid>,
+
+        // The keys to be used for this collection. This may be the default key (ie,
+        // used for all collections) or one specific to the collection, but the
+        // engine doesn't need to know that.
+        keys: KeyBundle,
+
+        // Commands the engine should act upon before syncing.
+        commands: Vec<Command>
+    }
+
+     struct Command {
+        ...
+    }
+
+## APIs
+## Sync Manager
+
+    // Register an engine.
+    fn register_engine(name: String, collections: Vec<String>, sync_callback: Fn<...>)
+
+    // Get the list of engines which are declined. There's a bit of confusion here
+    // between an "engine" and a "collection" - eg, assuming history ends up using
+    // 2 collections, the UI will want 1 entry.
+    // Maybe we define this as a "collection" and the app should ignore collections
+    // it doesn't understand?
+    // Will return None if we haven't yet done a first sync so don't know. Will
+    // return an empty vec if we have synced but no engines are declined.
+    fn get_declined() -> Option<Vec<String>>
+
+    // Ditto here - a bit of confusion between "collection" and "engine" here.
+    fn enable_engine(String, bool)
+
+    // TODO: let's think a little more about how desktop uses a single "addons"
+    // pref which controls addons *and* storage.sync, and how a single "Forms"
+    // pref controls 2 discrete engines.
+
+    // Do a sync now. There's a special error code to indicate that something
+    // seems to have gone wrong with auth, which is surfaced to the app, so
+    // it can call on FxA to do magic.
+    fn sync_now(engines: Option<Vec<String>>) -> Result<()>
+
+    // Stop syncing ASAP. Called as the app is shutting down, and possibly
+    // in other circumstances, such as when wifi is lost but the user has
+    // opted in to syncing only on wifi
+    // TBD - does this make sense? If so, how long is the app expected to wait?
+    fn stop()
+
+    // others?
+
+## Engines
+
+    // Perform the sync of the engine. When asked to sync, the manager will
+    // create a GlobalEngineState specific to the engine and pass it in.
+    fn sync(state: GlobalEngineState) -> ???
```

Comment:
For Firefox iOS, they current have code to pass in a (tokenserver_uri, oauth_token) pair to the logins `sync` method.  It seems reasonable to expect them to update that code to instead pass in a synthesized `GlobalEngineState` object created out of the state of their existing sync infrastructure.

For Lockbox, will we require them to create and drive a new `SyncManager` object of some sort?  Or will we maintain a higher-level API that matches the current behaviour, taking an oauth token and doing the sync-manager logic internally?  There's a risk here that we'll end up splintering the consumer experience for Lockbox on Android (where they can use the a-c `SyncManager` plumbing) and Lockbox on iOS (where they may not have as many nice abstractions).

Reply:
> How does this `sync` method relate to the `sync` method on, say, the logins store that is already called by lockbox? Is it a replacement, or is it on a different object, or something else?

Both? :) I think this will be a replacement *and* on a different object. ie, lockbox will still have just 1 sync function to call, but it may obtain that function in a slightly different way. We can probably make this invisible to them though. I also hope that this will be the same on iOS.

I suspect that the general sync process will still take the same inputs as it does now, it's just that the responsibilities for consuming those inputs will shift a little. I'll try and be more specific about this in the next version.